File: frontend/src/services/announcement.ts
// src/services/announcement.ts
import api from './api';
import {
  Announcement,
  AnnouncementCreatePayload,
  AnnouncementUpdatePayload,
  AnnouncementListResponse,
} from '../types';

export interface AnnouncementParams {
  audience?: string;
  grade_id?: number;
  section_id?: number;
  status?: string;
  skip?: number;
  limit?: number;
}

export const announcementService = {
  async listAnnouncements(params?: AnnouncementParams): Promise<AnnouncementListResponse> {
    const response = await api.get<AnnouncementListResponse>('/announcements/', { params });
    return response.data;
  },

  async getAnnouncement(id: number): Promise<Announcement> {
    const response = await api.get<Announcement>(`/announcements/${id}`);
    return response.data;
  },

  async createAnnouncement(payload: AnnouncementCreatePayload): Promise<Announcement> {
    const response = await api.post<Announcement>('/announcements/', payload);
    return response.data;
  },

  async updateAnnouncement(id: number, payload: AnnouncementUpdatePayload): Promise<Announcement> {
    const response = await api.put<Announcement>(`/announcements/${id}`, payload);
    return response.data;
  },

  async deleteAnnouncement(id: number): Promise<Announcement> {
    const response = await api.delete<Announcement>(`/announcements/${id}`);
    return response.data;
  },

  async publishAnnouncement(id: number): Promise<Announcement> {
    const response = await api.post<Announcement>(`/announcements/${id}/publish`);
    return response.data;
  },

  async archiveAnnouncement(id: number): Promise<Announcement> {
    const response = await api.post<Announcement>(`/announcements/${id}/archive`);
    return response.data;
  },
};

File: backend-python/app/routers/v1/announcement.py
# app/routers/v1/announcement.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session
from app.api import deps
from app.schemas.announcement import (
    AnnouncementResponse,
    AnnouncementListResponse,
    AnnouncementCreate,
    AnnouncementUpdate,
    AnnouncementStatus,
    AnnouncementAudience,
)
from app.services.announcement import AnnouncementService
from app.models.user import UserModel

router = APIRouter(prefix="/announcements", tags=["Announcements"])


@router.get("/", response_model=AnnouncementListResponse)
def list_announcements(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    audience: Optional[AnnouncementAudience] = None,
    grade_id: Optional[int] = None,
    section_id: Optional[int] = None,
    status_filter: Optional[AnnouncementStatus] = Query(None, alias="status"),
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_user),
):
    items, total = AnnouncementService.list_announcements(
        db,
        skip=skip,
        limit=limit,
        audience=audience,
        grade_id=grade_id,
        section_id=section_id,
        status=status_filter,
        current_user=current_user,
    )
    return {"total": total, "items": items}

@router.get("/{announcement_id}", response_model=AnnouncementResponse)
def get_announcement(
    announcement_id: int,
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_user),
):
    return AnnouncementService.get_announcement(db, announcement_id=announcement_id, current_user=current_user)

@router.post("/", response_model=AnnouncementResponse, status_code=status.HTTP_201_CREATED)
def create_announcement(
    obj_in: AnnouncementCreate,
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_teacher),
):
    return AnnouncementService.create_announcement(db, obj_in=obj_in, current_user=current_user)

@router.put("/{announcement_id}", response_model=AnnouncementResponse)
def update_announcement(
    announcement_id: int,
    obj_in: AnnouncementUpdate,
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_user),
):
    return AnnouncementService.update_announcement(db, announcement_id=announcement_id, obj_in=obj_in, current_user=current_user)

@router.delete("/{announcement_id}", response_model=AnnouncementResponse)
def delete_announcement(
    announcement_id: int,
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_user),
):
    return AnnouncementService.delete_announcement(db, announcement_id=announcement_id, current_user=current_user)

@router.post("/{announcement_id}/publish", response_model=AnnouncementResponse)
def publish_announcement(
    announcement_id: int,
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_principal),
):
    return AnnouncementService.publish_announcement(db, announcement_id=announcement_id, current_user=current_user)

@router.post("/{announcement_id}/archive", response_model=AnnouncementResponse)
def archive_announcement(
    announcement_id: int,
    db: Session = Depends(deps.get_db),
    current_user: UserModel = Depends(deps.get_current_active_principal),
):
    return AnnouncementService.archive_announcement(db, announcement_id=announcement_id, current_user=current_user)